Updated exposure function call: enable temporal filtering

erialC-P opened the pull request:
This PR updates the exposure function to incorporate the ability to include temporal filtering in the exposure output/s. Incorporating this update into the intertidal repo is required to enable moving the exposure demo notebook into the dea_notebooks repository whilst ensuring enabled functionality is provided by the intertidal repository.

Changes and updates:
- all updates to the function call default to `None` so theoretically, no updates should be required to the remainder of our production code base.
- [Important] returned exposure values are now included in a dict object. A working prototype for handling this is included in the `Intertidal_workflow.ipynb`. 
- Optional inclusions in this new exposure function include the ability to apply temporal filters to the exposure high temporal resolution tidal modelling (e.g. winter dates only). Separately, tuples of filters can be applied (e.g. winter daytime). Lastly, functionality is built in to return the tide modelling and filtered tide height values to enable demonstration of the exposure method in other workflows.

Future work:
- combine single and tuple filter options into single function call
- thoroughly check all docstrings to ensure they are up to date

erialC-P committed:
removed spatial filtering from exposure function and updated function call in notebook

erialC-P committed:
Updated to be consistent with closed PR #85

Removed a outstanding references to spatial filtering and ensured script is up to date and consistent with closed PR #85

erialC-P committed:
updated cli to use new exposure func outputs

erialC-P committed:
Automatically update integration test validation results

robbibt commented on intertidal/elevation.py:
```diff
@@ -1264,22 +1264,24 @@ def intertidal_cli(
             )
+            
+            # Write each exposure output as new variables in the main dataset
+            for x in exposure_filters.data_vars:
```
I think given we're only ever going to be extracting the "unfiltered" layer here, I think we can simplify this code to:
```
ds["exposure"] = exposure_filters["unfiltered"]
```

robbibt commented on intertidal/exposure.py:
```diff
@@ -1,35 +1,286 @@
+    For an 'unfiltered', all of epoch-time, analysis, exposure is
+    calculated per pixel. All other filter options calculate exposure
+    from high temporal resolution modelled tides that are averaged 
+    across the nominated area of interest only.
```
```suggestion
    into a 1D timeseries across the nominated area of interest.
```

robbibt commented on intertidal/exposure.py:
```diff
@@ -80,31 +389,163 @@ def exposure(
-    # Calculate the tide-height difference between the elevation value and
-    # each percentile value per pixel
-    diff = abs(tide_cq - dem)
+    modelledtides_1d = modelledtides_lowres.mean(dim=["x", "y"])
```
There's a similar code block below this that is commented out - can you have a look and see if we can remove one of these?

robbibt commented on intertidal/exposure.py:
```diff
@@ -80,31 +389,163 @@ def exposure(
+            .tide_m
+        )
+
+        # Add modelledtides_x to output dict
```
```suggestion
        # Add modelledtides_x to output dataset
```

robbibt commented on intertidal/exposure.py:
```diff
@@ -80,31 +389,163 @@ def exposure(
+            z = x[1]
+            timeranges[str(y + "_" + z)] = timeranges[y].intersection(timeranges[z])
+
+    # Intersect datetimes of interest with the low-res tidal model
```
```suggestion
    # Intersect datetimes of interest with the 1D tidal model
```

robbibt commented on intertidal/elevation.py:
```diff
@@ -1264,22 +1264,24 @@ def intertidal_cli(
-
             # Calculate exposure
-            ds["exposure"], tide_cq = exposure(
+            exposure_filters, tide_cq_dict = exposure(
```
In the function below, `tide_cq_dict` is called `modelledtides_ds` - can we use that terminology here too for consistency? 

(are these raw modelled tides or tide quantiles? Alternatively, we could call it something like `tide_quantiles_ds` too if that was more clear)

erialC-P replied:
I'll update that variable name. It's a legacy hangover from an earlier version of the code.

erialC-P committed:
Update intertidal/exposure.py

Co-authored-by: Robbi Bishop-Taylor <Robbi.BishopTaylor@ga.gov.au>

erialC-P committed:
Update intertidal/exposure.py

Co-authored-by: Robbi Bishop-Taylor <Robbi.BishopTaylor@ga.gov.au>

erialC-P committed:
Update intertidal/exposure.py

Co-authored-by: Robbi Bishop-Taylor <Robbi.BishopTaylor@ga.gov.au>

erialC-P committed:
Automatically update integration test validation results

erialC-P committed:
pre-git pull updates to add PR review changes

erialC-P committed:
PR #88 reviews

updated:
- library import statements
- docstrings
- reviewer comments from PR #88

erialC-P committed:
PR #88 reviews

- applied Black formatting
- updated exposure func call in line with returned function variable names
- simplified the addition of 'unfiltered exposure' into master ds

erialC-P committed:
Black formatting

erialC-P committed:
PR #88 review

updated exposure function call to align with returned variable names in function definition

erialC-P committed:
bug fix - separate unfiltered calculation from filtered calculation

erialC-P committed:
bug fix - updated variable naming for consistency

erialC-P committed:
Automatically update integration test validation results

robbibt reviewed (approved):
Wooo!
